File: Projects-Solutions/Numbers/Go/factorial.go
/*Factorial Finder** - The Factorial of a positive integer, n,
* is defined as the product of the sequence n, n-1, n-2, ...1 and the factorial of zero, 0,
* is defined as being 1. Solve this using both loops and recursion.
 */

package main

import "fmt"

/* factorial by recursion*/
func fact_recursion(n uint) uint {
	if n == 0 || n == 1 {
		return 1
	} else {
		return n * fact_recursion(n-1)
	}
}

/* factorial by loop*/
func fact_loop(n uint) uint {
	var factorial uint = 1

	for ; n > 0; n-- {
		factorial *= n
	}

	return factorial
}

func main() {
	var num uint
	fmt.Print("Enter number find it's factorial: ")
	_, err := fmt.Scanf("%d", &num)

	if err != nil {
		panic(err)
	}

	if num >= 0 && num <= 65 {
		fmt.Printf("Factorial of %d by recursion is %d\n", num, fact_recursion(num))
		fmt.Printf("Factorial of %d by loops is %d\n", num, fact_loop(num))
	} else {
		fmt.Println("Limits : 0 >= num <= 65")
	}
}

File: Projects-Solutions/Numbers/Java/Factorial.java
/*Factorial Finder :
* The Factorial of a positive integer, n, is defined as the product of the sequence n, n-1, n-2, ...1 
* and the factorial of zero, 0, is defined as being 1. Solve this using both loops and recursion.
*/

import java.math.BigInteger;
import java.util.Scanner;

public class Factorial {
	//factorial by recursion
	public static BigInteger factorial(int n){
		if(BigInteger.valueOf(n).equals(BigInteger.ONE) || BigInteger.valueOf(n).equals(BigInteger.ZERO)){
			return BigInteger.ONE;
		}else{
			return BigInteger.valueOf(n).multiply(factorial(n - 1));
		}
	}
	
	public static void main(String[] args){
		Scanner input = new Scanner(System.in);
		BigInteger fact = BigInteger.ONE;
		
		System.out.println("Enter an integer");
		int num = input.nextInt();
		
		if(num < 0){
			System.out.println("Input should be a positive integer\nExiting...");
			System.exit(1);
		}
		
		//factorial using a loop
		for(int i = num; i > 0; i--){
			fact = fact.multiply(BigInteger.valueOf(i));
		}
		
		System.out.println("\nUsing a loop, the factorial is: ");
		System.out.println(fact);
		
		System.out.println("\nBy recursion, the factorial is: ");
		System.out.println(factorial(num));
	}
}
